io/include/pcl/io/real_sense/librealsense/real_sense_device_manager.h

```diff
@@ -0,0 +1,164 @@
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef PCL_IO_REAL_SENSE_DEVICE_MANAGER_H
+#define PCL_IO_REAL_SENSE_DEVICE_MANAGER_H
+
+#include <boost/utility.hpp>
+#include <boost/shared_ptr.hpp> 
+#include <boost/weak_ptr.hpp>
+#include <boost/thread/mutex.hpp>
+#include <vector>
+
+#include <pcl/pcl_exports.h>
+
+#include <librealsense/rs.hpp>
+
+namespace pcl
+{
+
+  class RealSenseGrabber;
+
+  namespace io
+  {
+
+    namespace real_sense
+    {
+
+      class RealSenseDevice;
+
+      class PCL_EXPORTS RealSenseDeviceManager : boost::noncopyable
+      {
+        
+        public:
+
+          typedef boost::shared_ptr<RealSenseDeviceManager> Ptr;
+
+          static Ptr&
+          getInstance ()
+          {
+            static Ptr instance;
+            if (!instance)
+            {
+              boost::mutex::scoped_lock lock (mutex_);
+              if (!instance)
+                instance.reset (new RealSenseDeviceManager);
+            }
+            return (instance);
+          }
+
+          inline size_t
+          getNumDevices ()
+          {
+            return (device_list_.size ());
+          }
+
+          boost::shared_ptr<RealSenseDevice>
+          captureDevice ();
+
+          boost::shared_ptr<RealSenseDevice>
+          captureDevice (size_t index);
+
+          boost::shared_ptr<RealSenseDevice>
+          captureDevice (const std::string& sn);
+
+          ~RealSenseDeviceManager ();
+
+        private:
+
+          struct DeviceInfo
+          {
+            int index;
+            std::string serial;
+            bool isCaptured;
+          };
+
+          /** If the device is already captured returns a pointer. */
+          boost::shared_ptr<RealSenseDevice>
+          capture (DeviceInfo& device_info);
+
+          RealSenseDeviceManager ();
+
+          rs::context context_;
+
+          /** This function discovers devices that are capable of streaming
+            * depth data. */
+          void
+          populateDeviceList ();
+
+          std::vector<DeviceInfo> device_list_;
+
+          static boost::mutex mutex_;
+      };
+
+      class PCL_EXPORTS RealSenseDevice : boost::noncopyable
+      {
+        public:
+
+          typedef boost::shared_ptr<RealSenseDevice> Ptr;
+
+          inline const std::string&
+          getSerialNumber () { return (device_id_); }
+
+          inline rs::device*
+          getDevice () { return (device_); }
+
+          /** Reset the state of given device by releasing and capturing again. */
+          static void
+          reset (RealSenseDevice::Ptr& device)
+          {
+            std::string id = device->getSerialNumber ();
+            device.reset ();
+            device = RealSenseDeviceManager::getInstance ()->captureDevice (id);
+          }
+
+        private:
+
+          friend class RealSenseDeviceManager;
+
+          std::string device_id_;
+          rs::device* device_;
```

Comment:
I don't like this very much. You're passing a pointer to the user which is managed by the realsense. If the user wants he can delete the pointer. Instead I propose to hold a reference to the rs::device. The user can do whatever he wants with the reference without being able to delete the resource.